skrub/_multi_agg_joiner.py

```diff
@@ -0,0 +1,499 @@
+"""
+The MultiAggJoiner extends AggJoiner to multiple auxiliary tables.
+"""
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_is_fitted
+
+from skrub import _join_utils
+from skrub._agg_joiner import AggJoiner
+from skrub._dataframe._namespace import is_pandas, is_polars
+from skrub._utils import _is_array_like, atleast_2d_or_none
+
+
+class MultiAggJoiner(TransformerMixin, BaseEstimator):
+    """Extension of the AggJoiner to multiple auxiliary tables.
+
+    Apply numerical and categorical aggregation operations on the `cols`
+    to aggregate, selected by dtypes. See the list of supported operations
+    at the parameter `operations`.
+
+    If `cols` is not provided, `cols` is set to a list of lists.
+    For each table in `aux_tables`, the correponding list will be all columns
+    of that table, except the `aux_keys` associated with that table.
+
+    As opposed to the :class:`AggJoiner`, here `aux_tables` is a list of tables,
+    each of which will be joined on the main table. Therefore `aux_keys` is now
+    a list of keys, of the same length as `aux_tables`, and each entry in `aux_keys`
+    is used to join the corresponding aux table. In the same way, each entry in `cols`
+    is a list of columns to aggregate in the corresponding aux table.
+
+    Therefore if we have a single table, we could either use
+
+    - the AggJoiner: ``AggJoiner(aux_table, key="ID")``
+    - or the MultiAggJoiner: ``MultiAggJoiner([aux_table], keys=["ID"])``
+
+    Note that for `keys`, `main_keys`, `aux_keys`, `cols` and `operations`,
+    an input of the form ``[["a"], ["b"], ["c", "d"]]`` is valid
+    while ``["a", "b", ["c", "d"]]`` is not.
+
+    Using a column from the first aux table to join the second aux table is not
+    (yet) supported.
+
+    Accepts :obj:`pandas.DataFrame` and :class:`polars.DataFrame` inputs.
+
+    Parameters
+    ----------
+    aux_tables : iterable of DataFrameLike or "X"
+        Auxiliary dataframes to aggregate then join on the base table.
+        The placeholder string "X" can be provided to perform
+        self-aggregation on the input data. To provide a single auxiliary table,
+        ``aux_tables = [table]`` is supported, but not ``aux_tables = table``.
+        It's possible to provide both the placeholder "X" and auxiliary tables,
+        as in ``aux_tables = [table, "X"]``. If that's the case, the second table will
+        be replaced by the input data.
+
+    keys : iterable of str, or iterable of iterable of str, default=None
```

Comment:
In some way, I wondering if by allowing a single iterable, a user could start to think that he can do single agg-join.

Reply:
yes I would prefer having only one option. but if others want to avoid typing the [] for single-item lists I'm ok with that, too. at least now that AggJoiner and MultiAggJoiner are separate things we have 2 options not 3 :sweat_smile:

Reply:
The downside is that only accepting `iterable of iterable of str` makes a clunky API for single table joins. I guess in that case we enforce that people use the right estimator i.e. the `AggJoiner`

Reply:
If that's the case, we could remove the example in the docstring where we convert the `AggJoiner` example into a `MultiAggJoiner` example make make it more explicit in the docstring

Reply:
in any case 1 list of 10 strings would mean join 10 tables, each on 1 column right? and not join 1 table on 10 columns

but to avoid that ambiguity it might be good to force one type only , ie if I want to join 1 table on 1 column with the multiaggjoiner i need to pass `[ [ "my_column" ] ]`

Reply:
The `iterable of str` case only works for joining a single table, to join multiple tables, an `iterable of iterable of str` is needed, i.e.:
- `["a", "b", "c", "d"]` would join 1 table on 4 columns
- `[["a"], ["b"], ["c"], ["d"]]` would join 4 tables on 1 column each

This was done to avoid having to pass `[ [ "my_column" ] ]`, but in this case the user might as well use the `AggJoiner` -> changing it to enforce one type and avoid any confusion